src/Npgsql/Replication/PgOutput/NpgsqlPgOutputConnectionExtensions.cs

```diff
@@ -0,0 +1,92 @@
+        /// and <a href="https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html">https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html</a>
+        /// for more information.
+        /// </remarks>
+        /// <param name="connection">The <see cref="NpgsqlLogicalReplicationConnection"/> to use for creating the replication slot</param>
+        /// <param name="slotName">The name of the slot to create. Must be a valid replication slot name (see
+        /// <a href="https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS-MANIPULATION">https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS-MANIPULATION</a>).
+        /// </param>
+        /// <param name="temporarySlot">
+        /// <see langword="true"/> if this replication slot shall be temporary one; otherwise <see langword="false"/>.
+        /// Temporary slots are not saved to disk and are automatically dropped on error or when the session has finished.
+        /// </param>
+        /// <param name="slotSnapshotInitMode">
+        /// A <see cref="NpgsqlLogicalSlotSnapshotInitMode"/> to specify what to do with the snapshot created during logical slot
+        /// initialization. <see cref="NpgsqlLogicalSlotSnapshotInitMode.Export"/>, which is also the default, will export the
+        /// snapshot for use in other sessions. This option can't be used inside a transaction.
+        /// <see cref="NpgsqlLogicalSlotSnapshotInitMode.Use"/> will use the snapshot for the current transaction executing the
+        /// command. This option must be used in a transaction, and <see cref="NpgsqlLogicalSlotSnapshotInitMode.Use"/> must be the
+        /// first command run in that transaction. Finally, <see cref="NpgsqlLogicalSlotSnapshotInitMode.NoExport"/> will just use
+        /// the snapshot for logical decoding as normal but won't do anything else with it.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.
+        /// </param>
+        /// <returns>
+        /// A <see cref="NpgsqlLogicalReplicationSlot"/> that wraps the newly-created replication slot.
+        /// </returns>
+        public static async Task<NpgsqlPgOutputReplicationSlot> CreatePgOutputReplicationSlot(
+            this NpgsqlLogicalReplicationConnection connection,
+            string slotName,
+            bool temporarySlot = false,
+            NpgsqlLogicalSlotSnapshotInitMode? slotSnapshotInitMode = null,
+            CancellationToken cancellationToken = default)
+        {
+            // We don't enter NoSynchronizationContextScope here since we (have to) do it in CreateReplicationSlotForPlugin, because
```

Comment:
Note that this isn't correct - each method is responsible for itself for avoiding the user's sync context. In fact, as the code currently is, the continuation in this method (the `return` line) will run in the user's synchronization context - which isn't good.

Ordinarily every method must use ConfigureAwait(false) on every single await; we have our special NoSynchronizationContextScope trick, which only takes effect on methods called from within it - but not for methods above it.

Reply:
Where you're correct, is that since CreateReplicationSlotForPlugin is publicly-exposed (to plugins), it must indeed apply NoSynchronizationContextScope within it (since external plugin writers won't - it's an internal Npgsql thing).

But this method then must also do ConfigureAwait(false) for itself (or use NoSynchronizationContextScope - though that's useless).

Reply:
https://github.com/Brar/npgsql/pull/8